### DISCUSSION TITLE
Custom Script - stuck at generating URL to specific id (Device,IPAddress, ..)

### DISCUSSION THREAD

**Discussion by carsten-AEI** [netbox-community/netbox#13527] [Help Wanted!] [answered]
Hi,

sorry for bothering with this, but I'm currently stuck writing some hopefully helpful URLs for a simple custom script which should allow me to quickly rename a device and all related  `dns_name` entries. The script itself works (see the very bottom for reference), but as I usually also need to change the related IP addresses manually and would like to link to them directly in the output.

However, while I get the `url` attribute when I query via cURL I don't seem to get this information directly within the custom script. I've seen https://github.com/netbox-community/netbox/discussions/7054 and others but I'm too much of a Python n00b to be able to use the given advice.

When looking at `self.request` I could try to jumble together the API URL form `HTTP_REFERER` but I guess this would be quite error prone. `vrs(addr)` only yields something like
```
'_state': <django.db.models.base.ModelState object at 0x7f18770d0190>,
 'address': IPNetwork('1.2.3.4/32'),
 'assigned_object_id': 10178,
 'assigned_object_type_id': 5,
 'comments': '',
 'created': datetime.datetime(2021, 2, 26, 0, 0, tzinfo=datetime.timezone.utc),
 'custom_field_data': {'IP comment': None},
 'description': '',
 'dns_name': 'changed_dns_name',
 'id': 8372,
 'last_updated': datetime.datetime(2023, 8, 23, 6, 38, 2, 904050, tzinfo=datetime.timezone.utc),
 'nat_inside_id': None,
 'role': '',
 'status': 'dhcp',
 'tenant_id': 1,
 'vrf_id': None}
```

and here I'm stuck. Anyone who could give me a nudge how to proceed?

Thanks a lot in advance!

```
class ChangeDeviceName(Script):
    class Meta:
        name = "Rename device"
        description = "Change name of device and related interfaces"
        commit_default = False

    cur_name = StringVar(label="Current device name")
    new_name = StringVar(label="Current device name")
    update_dns = BooleanVar(label="Update dns names?", default=True)

    def run(self, data, commit):
        # some checks: current name should be unique and new name should not be taken yet
        dev = Device.objects.get(name=str(data['cur_name']))
        tmp = Device.objects.filter(name=str(data['new_name']))
        if len(tmp) > 0:
            return self.log_failure(f"Device name {data['new_name']} seems to exist already!")

        output = [f"Found device '{dev.name}' with id {dev.id}"]

        # change name of dev
        dev.name = str(data['new_name'])
        dev.save()

        if data['update_dns']:
            pattern = re.compile(str(data['cur_name']))
            # iterate over all related interfaces
            ifaces = dev.interfaces.all()
            for addr in IPAddress.objects.filter(interface__in=ifaces):
                old = addr.dns_name
                if pattern.search(old):
                    addr.dns_name = re.sub(pattern, str(data['new_name']), addr.dns_name, count=1)
                    output.append(f"Changed {old} to {addr.dns_name}")
                    addr.save()

        return '\n'.join(output)
```

**Comment by candlerb** [comment, marked as answer]
> However, while I get the `url` attribute when I query via cURL I don't seem to get this information directly within the custom script.

The Netbox REST API, and the Django ORM used within custom scripts, are two completely separate things.

Netbox provides a helper [get_absolute_url](https://docs.djangoproject.com/en/4.2/ref/models/instances/#get-absolute-url) to turn a model instance into its URL, although actually it's just a relative URL. For example, this is what I see in nbshell:

```
/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
### NetBox interactive shell (netbox3)
### Python 3.8.10 | Django 4.1.10 | NetBox 3.5.8
### lsmodels() will show available models. Use help(<model>) for more info.
>>> ip = IPAddress.objects.first()
>>> ip.get_absolute_url()
'/ipam/ip-addresses/467/'
>>>
```

If you want an absolute URL, I believe you'll need to prepend the base yourself - e.g. by wrapping it in `request.build_absolute_uri(...)`. (See how `dcim/svg/racks.py` and `dcim/api/views.py` does this for absolute links in rack elevation SVGs)

Note that if you use the [logging functions](https://docs.netbox.dev/en/stable/customization/custom-scripts/#logging) within a custom script, these support Markdown rendering which means the log messages can easily contain clickable links. For example:

```
self.log_success(f"Created VM [{vm.name}](/virtualization/virtual-machines/{vm.id}/)")
```

**Reply by carsten-AEI** [reply]
Thanks a lot!  This got me kickstarted, especially the hint with `get_absolute_url` and that ORM and REST API are quite distinct - I was not aware!

For completeness, the script now looks like this and seems to do what I want it to do :)
```
import re
from dcim.models import Device
from ipam.models import IPAddress
from extras.scripts import Script, StringVar, BooleanVar
from urllib.parse import urlparse

class ChangeDeviceName(Script):
    class Meta:
        name = "Rename device"
        description = "Change name of device and related interfaces"
        commit_default = False
        scheduling_enabled = False
        job_timeout = 10

    cur_name = StringVar(label="Current device name")
    new_name = StringVar(label="Current device name")
    update_dns = BooleanVar(label="Update dns names?", default=True)

    def run(self, data, commit):
        # extract calling URL from HTTP_REFERER
        parsed_refer = urlparse(self.request.META.get('HTTP_REFERER'))
        server = f"{parsed_refer.scheme}://{parsed_refer.netloc}"

        # some checks: current name should be unique and new name should not be taken yet
        dev = Device.objects.get(name=str(data['cur_name']))
        tmp = Device.objects.filter(name=str(data['new_name']))
        if len(tmp) > 0:
            return self.log_failure(f"Device name {data['new_name']} seems to exist already!")

        output = [f"Found current device [**`{dev.name}`**]({server}{dev.get_absolute_url()})."]
        # change name of dev
        if dev.pk and hasattr(dev, 'snapshot'):
            dev.snapshot()

        dev.name = str(data['new_name'])
        dev.full_clean()
        dev.save()

        if data['update_dns']:
            # iterate over interfaces
            pattern = re.compile(str(data['cur_name']))
            ifaces = dev.interfaces.all()
            for addr in IPAddress.objects.filter(interface__in=ifaces):
                old = addr.dns_name
                if pattern.search(old):
                    if addr.pk and hasattr(addr, 'snapshot'):
                        addr.snapshot()
                    addr.dns_name = re.sub(pattern, str(data['new_name']), addr.dns_name, count=1)
                    output.append(f"1. Changed **`{old}`** to **`{addr.dns_name}`**, [{addr}]({server}{addr.get_absolute_url()})")
                    addr.full_clean()
                    addr.save()

        self.log_success('\n'.join(output))
```

**Reply by candlerb** [reply]
Glad it's working.

>        parsed_refer = urlparse(self.request.META.get('HTTP_REFERER'))
>        server = f"{parsed_refer.scheme}://{parsed_refer.netloc}"

I haven't tried it, but does `self.request.build_absolute_uri('/')` work here? If so, you can trim the trailing slash.

I am dubious about using the `Referer` header, since it depends on which web page you viewed *before* clicking this one. The `Host` header is more likely what you want.

**Reply by carsten-AEI** [reply]
I hope I don't miss the mark, but `self.request` seems to be of type `NetBoxFakeRequest` and thus down not seem to have `build_absolute_uri`.

`self.request.META` looks like this:
```
{'CONTENT_LENGTH': '723', 
'CONTENT_TYPE': 'multipart/form-data; boundary=---------------------------337818233316841972232021781354', 
'HTTP_ACCEPT': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,/;q=0.8', 
'HTTP_ACCEPT_ENCODING': 'gzip, deflate, br', 
'HTTP_ACCEPT_LANGUAGE': 'en-US,en;q=0.7,de;q=0.3', '
HTTP_HOST': '10.20.60.115:8001', 
'HTTP_REFERER': 'https://172.29.50.115/netbox/extras/scripts/change_name/ChangeDeviceName/', 
'HTTP_USER_AGENT': 'Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0',
'QUERY_STRING': '', '
REMOTE_ADDR': '10.20.60.115', 
'REQUEST_METHOD': 'POST', 
'SERVER_NAME': '10.20.60.115', 
'SERVER_PORT': '8001'}
```
and it contains quite a jumble of stuff. The `netbox` is running on a multi-homed host (`10.20.60.115` and `172.29.50.115`) and reverse proxied on that same host, i.e. `HTTP_HOST` would result in a URL on the wrong network.

Outside of `.META` I did not find anything "useful" which would allow me to construct a correct link. But as outlined at the beginning I'm far out of my depth here :wink:

**Reply by candlerb** [reply]
`HTTP_HOST` should give the hostname from the original incoming request - but in your case it doesn't, instead it's showing the address that the frontend proxy is connecting to.

You need to ensure the frontend proxy is passing on the Host header from the original request: for example, Apache has a setting `ProxyPreserveHost On`, as shown in the [sample configs](https://github.com/netbox-community/netbox/tree/master/contrib).

If you don't do this then you'll find other problems. Most commonly, when looking at Rack Elevations and you click on a device, you'll get the wrong URL.

**Reply by carsten-AEI** [reply]
Interesting! On this test box we use nginx as a reverse proxy and it was almost exactly set up as [nginx.conf](https://github.com/netbox-community/netbox/blob/master/contrib/nginx.conf) with modifications for `server_name` and `proxy_pass` as I'm currently using `http://127.0.0.1:8001/netbox` instead of `http://127.0.0.1:8001`.

But still, `HTTP_HOST` only showed `10.20.60.115:8001` initially (debug change yesterday) and `127.0.0.1:8001` after changing it in `proxy_pass` for nginx and `bind` for gunicorn.

`HTTP_HOST` only started to show the caller IP when I also added `proxy_set_header Host $host;` to `nginx.conf`. Do I have a weird border line configuration running here or should this also be added into the sample config upstream? Currently, the test VM runs Debian 11 with nginx 1.18.0-6.1+deb11u3.

**Reply by candlerb** [reply]
From the sample config, you should have had

```
        proxy_set_header X-Forwarded-Host $http_host;
```

and this should have shown up as `HTTP_X_FORWARDED_HOST` in the request.  When building URLs, Django will use this header (if it's present) in preference to the Host: header.

If you did have this line, then I don't know why it was missing from `self.request.META`.

**Reply by carsten-AEI** [reply]
this is odd, with this `nginx.conf`

```
server {
    listen [::]:443 ssl ipv6only=off;

    server_name netbox-test;

    ssl_certificate /etc/ssl/certs/netbox.crt;
    ssl_certificate_key /etc/ssl/private/netbox.key;

    client_max_body_size 25m;

    location /netbox/static/ {
        alias /opt/netbox/netbox/static/;
    }

    location /netbox {
        proxy_pass http://127.0.0.1:8001/netbox;
        proxy_set_header X-Forwarded-Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;
    }

    rewrite ^/$ /netbox permanent;
}

server {
    listen [::]:80 ipv6only=off;
    server_name _;

    # Redirect all other HTTP traffic to HTTPS
    location / {
        return 301 https://$host$request_uri;
    }
}
```

and a full restart (`systemctl restart netbox netbox-rq nginx`), then `dict(self.request.META)` in the Custom Script returns this

```
{'CONTENT_LENGTH': '723', 'CONTENT_TYPE': 'multipart/form-data; boundary=---------------------------166552860123068484972368470446', 'HTTP_ACCEPT': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,/;q=0.8', 'HTTP_ACCEPT_ENCODING': 'gzip, deflate, br', 'HTTP_ACCEPT_LANGUAGE': 'en-US,en;q=0.7,de;q=0.3', 'HTTP_HOST': '127.0.0.1:8001', 'HTTP_REFERER': 'https://172.29.50.115/netbox/extras/scripts/change_name/ChangeDeviceName/', 'HTTP_USER_AGENT': 'Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0', 'QUERY_STRING': '', 'REMOTE_ADDR': '127.0.0.1', 'REQUEST_METHOD': 'POST', 'SERVER_NAME': '127.0.0.1', 'SERVER_PORT': '8001'}
```

This is weird. Just checked `gunicorn.conf` and `netbox/netbox/configuration.py` but I don't see anything in there which would stood out to me as odd.

not sure if important, `nbshell` returns
```
### NetBox interactive shell (netbox-test)
### Python 3.9.2 | Django 4.1.10 | NetBox 3.5.8
```

**Reply by candlerb** [reply]
I suggest you run `tcpdump -i lo -nn -s0 -A tcp port 8001` to confirm whether the `X-Forwarded-(Host|Proto)` headers are being sent to Netbox.

If you check the [documentation for custom scripts](https://docs.netbox.dev/en/stable/customization/custom-scripts/#accessing-request-data) you'll see the following example:

```
ip_address = self.request.META.get('HTTP_X_FORWARDED_FOR') or \
    self.request.META.get('REMOTE_ADDR')
```

This suggests to me that all the `X-` headers ought to be present.

If tcpdump shows they are present, but they don't reach your script, then they are being stripped out somewhere - but I have no idea where.

**Reply by carsten-AEI** [reply]
yepp, both present:
```
[...]
POST /netbox/extras/scripts/change_name/ChangeDeviceName/ HTTP/1.0
X-Forwarded-Host: 172.29.50.115
X-Real-IP: ::ffff:172.31.254.0
X-Forwarded-Proto: https
Host: 127.0.0.1:8001
Connection: close
Content-Length: 723
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Language: en-US,en;q=0.7,de;q=0.3
Accept-Encoding: gzip, deflate, br
Referer: https://172.29.50.115/netbox/extras/scripts/change_name/ChangeDeviceName/
Content-Type: multipart/form-data; boundary=---------------------------390705613117421445642087380259
Origin: https://172.29.50.115
Cookie: csrftoken=TOKEN sessionid=SESSION
Upgrade-Insecure-Requests: 1
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: same-origin
Sec-Fetch-User: ?1
[..]
Server: gunicorn
Date: Thu, 24 Aug 2023 07:11:40 GMT
Connection: close
Content-Type: text/html; charset=utf-8
Location: /netbox/extras/scripts/results/131/
X-Request-ID: 398dd334-7320-49c7-a985-4bc0893cbe58
Vary: HX-Request, Cookie, origin
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
Cross-Origin-Opener-Policy: same-origin
X-Frame-Options: SAMEORIGIN
Content-Length: 0
[..]
GET /netbox/extras/scripts/results/131/ HTTP/1.0
X-Forwarded-Host: 172.29.50.115
X-Real-IP: ::ffff:172.31.254.0
X-Forwarded-Proto: https
Host: 127.0.0.1:8001
Connection: close
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Language: en-US,en;q=0.7,de;q=0.3
Accept-Encoding: gzip, deflate, br
Referer: https://172.29.50.115/netbox/extras/scripts/change_name/ChangeDeviceName/
Cookie: csrftoken=TOKEN; sessionid=SESSION
Upgrade-Insecure-Requests: 1
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: same-origin
Sec-Fetch-User: ?1
[...]
Server: gunicorn
Date: Thu, 24 Aug 2023 07:11:40 GMT
Connection: close
Content-Type: text/html; charset=utf-8
X-Request-ID: 335dd1c7-ec72-45ea-806f-191b8edf1855
Vary: HX-Request, Cookie, origin
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
Cross-Origin-Opener-Policy: same-origin
X-Frame-Options: SAMEORIGIN
Content-Length: 87125
Set-Cookie: csrftoken=TOKEN; expires=Thu, 22 Aug 2024 07:11:40 GMT; Max-Age=31449600; Path=/netbox; SameSite=Lax
[..]
<p>{'CONTENT_LENGTH': '723', 'CONTENT_TYPE': 'multipart/form-data; boundary=---------------------------390705613117421445642087380259', 'HTTP_ACCEPT': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,<em>/</em>;q=0.8', 'HTTP_ACCEPT_ENCODING': 'gzip, deflate, br', 'HTTP_ACCEPT_LANGUAGE': 'en-US,en;q=0.7,de;q=0.3', 'HTTP_HOST': '127.0.0.1:8001', 'HTTP_REFERER': 'https://172.29.50.115/netbox/extras/scripts/change_name/ChangeDeviceName/', 'HTTP_USER_AGENT': 'Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0', 'QUERY_STRING': '', 'REMOTE_ADDR': '127.0.0.1', 'REQUEST_METHOD': 'POST', 'SERVER_NAME': '127.0.0.1', 'SERVER_PORT': '8001'}</p></div></td>
```

Thus, no clue here either. Given that I have a work-around with the `nginx.conf` and seem to be the only one hitting this so far, shall we assume PEBKAC here and that I misconfigured something here?

**Reply by candlerb** [reply]
I think it's some filtering going on in Netbox, and that your particular use case hasn't been considered. Perhaps there's an allowlist of headers which end up in request.META, and some are missing.

**Reply by candlerb** [reply]
A quick grep turns up `HTTP_REQUEST_META_SAFE_COPY` in `netbox/utilities/constants.py`. You could try editing that and see what happens? If that solves your problem, you could raise a FR to increase the set of allowed headers. `HTTP_X_FORWARDED_PROTO` seems like an important one too, to generate `http://` or `https://` links correctly.

**Reply by carsten-AEI** [reply]
Sorry for the delay, I first needed to understand Python logging and how I could enabled some debug output from `copy_safe_request(request)` in `netbox/utilities/utils.py` to be useful.

With
```
diff --git a/netbox/utilities/constants.py b/netbox/utilities/constants.py
index 366d8f7..a0657f8 100644
--- a/netbox/utilities/constants.py
+++ b/netbox/utilities/constants.py
@@ -45,6 +45,8 @@ HTTP_REQUEST_META_SAFE_COPY = [
     'HTTP_REFERER',
     'HTTP_USER_AGENT',
     'HTTP_X_FORWARDED_FOR',
+    'HTTP_X_FORWARDED_HOST',
+    'HTTP_X_FORWARDED_PROTO',
     'QUERY_STRING',
     'REMOTE_ADDR',
     'REMOTE_HOST',
```
I finally get those extra bits into `META`:

```
{:CONTENT_LENGTH=>"723",
 :CONTENT_TYPE=>"multipart/form-data; boundary=---------------------------178440007540120566012032856803",
 :HTTP_ACCEPT=>"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,/;q=0.8",
 :HTTP_ACCEPT_ENCODING=>"gzip, deflate, br",
 :HTTP_ACCEPT_LANGUAGE=>"en-US,en;q=0.7,de;q=0.3",
 :HTTP_HOST=>"127.0.0.1:8001",
 :HTTP_REFERER=>"https://172.29.50.115/netbox/extras/scripts/change_name/ChangeDeviceName/",
 :HTTP_USER_AGENT=>"Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0",
 :HTTP_X_FORWARDED_HOST=>"172.29.50.115",
 :HTTP_X_FORWARDED_PROTO=>"https",
 :QUERY_STRING=>"",
 :REMOTE_ADDR=>"127.0.0.1",
 :REQUEST_METHOD=>"POST",
 :SERVER_NAME=>"127.0.0.1",
 :SERVER_PORT=>"8001"}
```

So, shall I try to formulate a FR with that?

**Reply by candlerb** [reply]
Yes, that sounds good to me. I suggest you add `HTTP_X_REAL_IP` to that list too, since that's another header that the [default nginx config](https://github.com/netbox-community/netbox/blob/master/contrib/nginx.conf) supplies.
